test-data/unit/check-protocols.test

```diff
@@ -0,0 +1,1827 @@
+
+[case testBasicCallableStructuralSubtyping]
+from typing import Callable, Generic, TypeVar
+
+def apply(f: Callable[[int], int], x: int) -> int:
+    return f(x)
+
+class Add5:
+    def __call__(self, x: int) -> int:
+        return x + 5
+
+apply(Add5(), 5)
+
+T = TypeVar('T')
+def apply_gen(f: Callable[[T], T]) -> T:
+    pass
+
+reveal_type(apply_gen(Add5())) # E: Revealed type is 'builtins.int*'
+def apply_str(f: Callable[[str], int], x: str) -> int:
+    return f(x)
+apply_str(Add5(), 'a') # E: Argument 1 to "apply_str" has incompatible type "Add5"; expected Callable[[str], int] \
+                       # N: 'Add5.__call__' has type 'Callable[[int], int]'
+[builtins fixtures/isinstancelist.pyi]
+
+[case testStructuralSupportForPartial]
+from typing import Callable, TypeVar, Generic, Any
+
+T = TypeVar('T')
+
+class partial(Generic[T]):
+    def __init__(self, func: Callable[..., T], *args: Any) -> None: ...
+    def __call__(self, *args: Any) -> T: ...
+
+def inc(a: int, temp: str) -> int:
+    pass
+
+def foo(f: Callable[[int], T]) -> T:
+    return f(1)
+
+reveal_type(foo(partial(inc, 'temp'))) # E: Revealed type is 'builtins.int*'
+[builtins fixtures/list.pyi]
+
+-- Standard protocol types (SupportsInt, Sized, etc.)
+-- --------------------------------------------------
+
+-- More tests could be added for types from typing converted to protocols
+
+[case testBasicSizedProtocol]
+from typing import Sized
+
+class Foo:
+    def __len__(self) -> int:
+        return 42
+
+def bar(a: Sized) -> int:
+    return a.__len__()
+
+bar(Foo())
+bar(1)
+
+[builtins fixtures/isinstancelist.pyi]
+[out]
+main:11: error: Argument 1 to "bar" has incompatible type "int"; expected "Sized"
+
+[case testBasicSupportsIntProtocol]
+from typing import SupportsInt
+
+class Bar:
+    def __int__(self):
+        return 1
+
+def foo(a: SupportsInt):
+    pass
+
+foo(Bar())
+foo('no way')
+
+[builtins fixtures/isinstancelist.pyi]
+[out]
+main:11: error: Argument 1 to "foo" has incompatible type "str"; expected "SupportsInt"
+
+-- Additional tests and corner cases for protocols
+-- ----------------------------------------------
+
+[case testAnyWithProtocols]
+from typing import Protocol, Any, TypeVar
+
+T = TypeVar('T')
+
+class P1(Protocol):
+    attr1: int
+class P2(Protocol[T]):
+    attr2: T
+class P3(Protocol):
+    attr: P3
+
+def f1(x: P1) -> None: pass
+def f2(x: P2[str]) -> None: pass
+def f3(x: P3) -> None: pass
+
+class C1:
+    attr1: Any
+class C2:
+    attr2: Any
+class C3:
+    attr: Any
+
+f1(C1())
+f2(C2())
+f3(C3())
+
+f2(C3())  # E: Argument 1 to "f2" has incompatible type "C3"; expected P2[str]
+a: Any
+f1(a)
+f2(a)
+f3(a)
+
+[case testErrorsForProtocolsInDifferentPlaces]
+from typing import Protocol
+
+class P(Protocol):
+    attr1: int
+    attr2: str
+    attr3: int
+
+class C:
+    attr1: str
+    @property
+    def attr2(self) -> int: pass
+
+x: P = C() # E: Incompatible types in assignment (expression has type "C", variable has type "P") \
```

Comment:
Do you test anywhere what happens if a class has method and the protocol has an attribute/property, and vice versa?